# huva (R)

In the last decades, the spread of increasingly affordable transcriptomic techniques and the concomitant production of population-scale multi-layered datasets assembled extensive experimental data of different types on single healthy individuals. Within this population, most genetic variation and environmental factors are influencing gene expression with no no clinical evidence of pathological states. Variance of gene expression, which is a characteristic of any given natural population, can be exploited to set up a conditional quasi loss- and gain-of-function “in population” experiment if expression values for the gene of interest (GOI) are available in a sufficiently large cohort. We describe here a novel tool, called huva (human variation), which takes advantage of population-scale multi-layer data to infer gene function and relationship between phenotype and gene expression. Within a reference dataset, huva derives two experimental groups, i.e. individuals with "low" or "high" expression of the GOI, enabling the subsequent comparison of their transcriptional profile and functional parameters. We demonstrate that this approach robustly and efficiently identifies the phenotypic relevance of a GOI allowing the stratification of genes according to shared functions. Additionally, we describe how huva can predict the phenotype of naturally occurring gain-of-function mutations in humans, further validating the biological power of the tool.